webapp/src/webapp/features/runbooks/runner/events.cljs

```diff
@@ -73,28 +79,271 @@
+                                (on-failure _error-message error)))
+         default-on-success (fn [res]
+                              (rf/dispatch
+                               [:show-snackbar {:level :success
+                                                :text "Runbook was executed!"}])
+                              (rf/dispatch [:runbooks/exec-success res file-name])
+                              (rf/dispatch [:webapp.events.editor-plugin/editor-plugin->set-script-success res file-name])
+                              (when on-success
+                                (on-success res)))
+         base-db (assoc db :runbooks->exec {:status :loading :data nil})]
+     (merge {:db base-db
+             :fx [[:dispatch [:fetch {:method "POST"
+                                      :uri "/runbooks/exec"
+                                      :on-success default-on-success
+                                      :on-failure default-on-failure
+                                      :body payload}]]]}
+            (when-not keep-metadata?
+              {:db (-> base-db
+                       (assoc-in [:runbooks :metadata] [])
+                       (assoc-in [:runbooks :metadata-key] "")
+                       (assoc-in [:runbooks :metadata-value] ""))})))))
+
+(rf/reg-event-fx
+ :runbooks/exec-success
+ (fn [{:keys [db]} [_ data file-name]]
+   {:db (assoc-in db [:runbooks->exec] {:status :success
+                                        :data (merge data {:file-name file-name})})}))
+
+(rf/reg-event-fx
+ :runbooks/exec-failure
+ (fn [{:keys [db]} [_ error]]
+   {:db (assoc-in db [:runbooks->exec] {:status :failure :data error})}))
+
+;; Execution Subscriptions
+(rf/reg-sub
+ :runbooks->exec
+ (fn [db]
+   (get-in db [:runbooks->exec] {:status :idle :data nil})))
+
+;; Metadata Events
 (rf/reg-event-db
- :runbooks/trigger-execute
- (fn [db _]
-   (assoc-in db [:runbooks :execute-trigger] true)))
+ :runbooks/add-metadata
+ (fn [db [_ metadata]]
+   (let [old-metadata (or (get-in db [:runbooks :metadata]) [])]
+     (update db :runbooks merge {:metadata (conj old-metadata metadata)
+                                 :metadata-key ""
+                                 :metadata-value ""}))))
+
+(rf/reg-event-db
+ :runbooks/update-metadata-key
+ (fn [db [_ value]]
+   (assoc-in db [:runbooks :metadata-key] value)))
 
 (rf/reg-event-db
- :runbooks/execute-handled
+ :runbooks/update-metadata-value
+ (fn [db [_ value]]
+   (assoc-in db [:runbooks :metadata-value] value)))
+
+(rf/reg-event-db
+ :runbooks/update-metadata-at-index
+ (fn [db [_ index field value]]
+   (assoc-in db [:runbooks :metadata index field] value)))
+
+;; Metadata Subscriptions
+(rf/reg-sub
+ :runbooks/metadata
+ (fn [db]
+   (get-in db [:runbooks :metadata] [])))
+
+(rf/reg-sub
+ :runbooks/metadata-key
+ (fn [db]
+   (get-in db [:runbooks :metadata-key] "")))
+
+(rf/reg-sub
+ :runbooks/metadata-value
+ (fn [db]
+   (get-in db [:runbooks :metadata-value] "")))
```

Comment:
question: why some many reg-sub for a map? does we need to look to change in those field ?

for e.g https://github.com/hoophq/hoop/pull/1125/files#diff-3d271c9dcbf988642f1dbe9eecf7b803343a87b95359dc3fb16f0f10858b9055R20
the header is looking for this field and if i am not wrong when any of those field change it will render the header..

so why not just (rf/reg-sub :runboks/runbook ...) => {:metadata [], :metadata-key "..."... } so on and have only 1 mapa and 1 sub ..

and when some events happens to update it will be render by the component that is looking for the mapa, because looks like the component importing is alwys looking for all three fields.


kind i dont have much context on this part of the code, so I might miss something..